altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -265,12 +269,70 @@ impl FreeList {
                 },
             }
         });
-        match previous {
-            Some(previous) => {
-                previous.next_block = alloc_block;
-                alloc_block.get_ref_mut().unwrap().next_block = Link::from(current);
+
+
+        // Determine if the deallocated block is adjacent to the leading free block.
+        let merge_with_previous = match previous {
```

Comment:
The deallocate function is getting kind of long as it is. It might make sense to break up all the node merging logic into a separate function, and call it at the end of deallocate.